### PR TITLE
Hide block toolbar when spacing visualizer is showing

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Hides the selected block toolbar when the padding/margin visualiser is showing.

## Why?
Part of the visualizer improvements from #44700

## How?
Adds new store action/reducer/selector to indicate when the spacing visualizer is showing, and uses the selector to show/hide the toolbar

## Testing Instructions

- Add a group block and adjust the margins and padding
- Make sure the block toolbar disappears when the visualizer appears

## Screenshots or screencast 

Before:

https://user-images.githubusercontent.com/3629020/196817894-4c39a498-4b70-4652-984d-567c23bba862.mp4

After:

https://user-images.githubusercontent.com/3629020/196817916-ea9e103e-6794-4d4c-847a-a1d048e4c970.mp4

### CODE REVIEW

**Comment by glendaviesnz** [comment]
@talldan the previous idea I mentioned of adding a `visualize` editor mode fell apart when also using the zoomed in mode, so have gone with a more targetted approach of adding a specific selector/reducer to handle when the visualizer is showing.

File: packages/block-editor/src/hooks/dimensions.js
Comment (talldan) [inline packages/block-editor/src/hooks/dimensions.js]: This is an interesting problem to solve.

From the point of a public API, if a plugin calls `startShowingVisualizer` it has the effect of hiding the block toolbar, which I think would be unexpected. I see why it's like that though—`startTyping` also works the same way. It does leave open the possibility of triggering other effects when a user starts or stops typing.

I'd question if it's needed though, and I think code like this might cause more confusion (e.g. I can imagine plugins calling `startTyping` to hide the toolbar when the user isn't actually typing, just to get the desired result). 

One option is that the actions/selectors/reducer can be more specific to what it does, e.g. `hideBlockToolbar` / `showBlockToolbar`.

That way there's the possibility of converging the `isTyping` code and the visualizer code to use the same system for hiding revealing the toolbar. Possibly deprecating `isTyping` along the way.

I think @ellatrix might be familiar with this code, so it'd be good to get a second opinion.

Comment (glendaviesnz) [reply packages/block-editor/src/hooks/dimensions.js]: yeh, as I was adding this I was thinking to myself "the naming isn't right", `hideBlockToolbar / showBlockToolbar` is much better from the external API point of view so have switched it to that.

**Comment by glendaviesnz** [comment]
@talldan because `isTyping` is used for a couple of other things besides hiding the block toolbar (useBlockClassnames && DownloadableBlocksPanel) I couldn't see a way that `hideBlockToolbar` could be used to deprecate `isTyping` completely.

We could have the `START_TYPING` action type trigger the `isBlockToolbarHidden` reducer as well as the `HIDE_BLOCK_TOOLBAR` action type so that in  `SelectedBlockPopover` we only have to check for `isBlockToolbarHidden` ... but not sure that having the `START_TYPING` action trigger the `isBlockToolbarHidden` reducer provides a much better API than the `isTyping` selector hiding the toolbar directly 🤔

**Comment by talldan** [comment]
It looks like `DownloadableBlocksPanel` doesn't use the `isTyping` state.

`useBlockClassnames` is interesting as it's used for hiding the block outline. In https://github.com/WordPress/gutenberg/pull/45056 I have the same need to hide the block outline as part of showing a visualizer component. 

Maybe `hideBlockToolbar` is too specific then, and it could be `hideBlockInterface` / `showBlockInterface`, that could also then be used to hide the block outline. 🤔

**Comment by glendaviesnz** [comment]
> It looks like DownloadableBlocksPanel doesn't use the isTyping state.

Ah, yeh, same name, different source.

> Maybe hideBlockToolbar is too specific then, and it could be hideBlockInterface / showBlockInterface, that could also
> then be used to hide the block outline. 🤔

I guess then the problem might be that in cases where you want to hide the outline you may not always want to hide the toolbar and vice versa ? 🤔

**Comment by talldan** [comment]
> I guess then the problem might be that in cases where you want to hide the outline you may not always want to hide the toolbar and vice versa ? 🤔

It's a difficult choice 🤔 . I don't think `showBlockInterface`/`hideBlockInterface` precludes this, as the action can accept additional parameters. However, right now this use case doesn't exist, and there are a few use cases for hiding or showing both the toolbar and the outline at the same time (margin/padding/alignment visualizers and typing).

The alternative is separate `isBlockToolbarHidden` and `isBlockOutlineHidden` state, but it might be too granular. If we merge this though, it'll be the only choice I have in solving for https://github.com/WordPress/gutenberg/pull/45056, so that's why I'm a little bit hesitant. 😄 

The main points for me are that it's undesirable having two APIs (`isTyping` and `isBlockToolbarHidden`) that do the same thing. However at the same time we know there are problems* trying to make them using a single bit of state. Whether that problem can be solved is probably the thing that determines what solution is best. I think it may be worth trying a few different things to solve that to see what works best.

* moving the mouse triggers `isTyping` and shows the block toolbar. This conflicts with the desired outcome for the visualizers, where the block toolbar is hidden while the user is moving the mouse.

**Comment by glendaviesnz** [comment]
@talldan I have changed it to `isBlockInterfaceHidden` so it can be used for your use case - as you not we can add additional params in the future if there ever arises a case where only specific parts of the interface should be hidden.

The difference in mouse actions between the current `isTyping` and the visualizer use case is tricky, I haven't been able to think of a way of combining them into a single API yet, but will keep thinking about it.

**Comment by glendaviesnz** [comment]
@talldan if we [stop propagation of the mouse event](https://github.com/WordPress/gutenberg/pull/45131/commits/ec9f990638e1e04c36c34d6400bcf38a76738455) in the context of the dimensions control then the `stopTyping` actions are not fired in my testing, so I think this would open the way for combining `isTyping` with `isBlockInterfaceHidden` 🤔 . Should we explore that in this PR, or look to get this merged as is and then explore that?

File: packages/block-editor/src/hooks/dimensions.js
Comment (talldan) [inline packages/block-editor/src/hooks/dimensions.js:64]: Are the `stopPropagation` calls still needed here?

**Review by talldan — APPROVED** [review]
This is a nice improvement, thanks for iterating on the code.

I had a few observations in testing, and they may or may not lead to small adjustments in how this works:
- When there's no padding or margin, the toolbar is still hidden, but there's no visualization, which doesn't feel quite right to me when using it. Not sure if this is something that's easy to fix.
- When changing a value and hovering off, the toolbar can reappear momentarily before the visualizer is hidden. I think that's because there's a timeout related to changing the value.
- Hiding the toolbar is possibly only needed for the margin visualizer, as the padding can't overlap the toolbar. I'm completely on the fence on this one, in some ways it's nice to have it work consistently and hide the toolbar in both cases. Not sure, what do you think?

I would consider these things as low priority though and they shouldn't block merge.

File: packages/block-editor/src/hooks/dimensions.js
Comment (glendaviesnz) [reply packages/block-editor/src/hooks/dimensions.js:64]: It is there to prevent these mouse events from triggering the stopTyping action in anticipation of us merging that with the showBlockInterface. I could remove them until we actually do that work, but thought it might be worth adding them to prevent this inadvertently breaking when we do get to it.

**Comment by glendaviesnz** [comment]
Thanks @talldan 

> When there's no padding or margin, the toolbar is still hidden, but there's no visualization, which doesn't feel quite right to me when using it. Not sure if this is something that's easy to fix.

I thought this would be easy but it opened a can of worms, partly related to the timeout mentioned below.

> When changing a value and hovering off, the toolbar can reappear momentarily before the visualizer is hidden. I think that's because there's a timeout related to changing the value.

This is related to keeping it working for the Select list when a theme has more than 7 spacing values. I tried fixing this, but combined with the option of not showing 0 values above turned out to cause a rabbit hole of problems 😄 

I am going to merge it as is so you can pick up those actions and will add some follow-up issues to deal with the above problems.

> Hiding the toolbar is possibly only needed for the margin visualizer, as the padding can't overlap the toolbar. I'm completely on the fence on this one, in some ways it's nice to have it work consistently and hide the toolbar in both cases. Not sure, what do you think?

Yeh, I wondered that, will leave as is and see what wider feedback is, easy to pull out of padding if the consensus is that it is not needed.

**Comment by talldan** [comment]
> I am going to merge it as is so you can pick up those actions and will add some follow-up issues to deal with the above problems.

Sounds good, thanks!

**Pull request merged**
